presto-common/src/main/java/com/facebook/presto/common/Page.java

```diff
@@ -0,0 +1,450 @@
+            }
+        }
+        // No newly loaded blocks
+        return this;
+    }
+
+    public Page getLoadedPage(int channel)
+    {
+        return wrapBlocksWithoutCopy(positionCount, new Block[] {this.blocks[channel].getLoadedBlock()});
+    }
+
+    public Page getLoadedPage(int... channels)
+    {
+        requireNonNull(channels, "channels is null");
+
+        Block[] blocks = new Block[channels.length];
+        for (int i = 0; i < channels.length; i++) {
+            blocks[i] = this.blocks[channels[i]].getLoadedBlock();
+        }
+        return wrapBlocksWithoutCopy(positionCount, blocks);
+    }
+
+    @Override
+    public String toString()
+    {
+        StringBuilder builder = new StringBuilder("Page{");
+        builder.append("positions=").append(positionCount);
+        builder.append(", channels=").append(getChannelCount());
+        builder.append('}');
+        builder.append("@").append(Integer.toHexString(System.identityHashCode(this)));
+        return builder.toString();
+    }
+
+    private static int determinePositionCount(Block... blocks)
+    {
+        requireNonNull(blocks, "blocks is null");
+        if (blocks.length == 0) {
+            throw new IllegalArgumentException("blocks is empty");
+        }
+
+        return blocks[0].getPositionCount();
+    }
+
+    public Page getPositions(int[] retainedPositions, int offset, int length)
+    {
+        requireNonNull(retainedPositions, "retainedPositions is null");
+
+        Block[] blocks = new Block[this.blocks.length];
+        for (int i = 0; i < blocks.length; i++) {
+            blocks[i] = this.blocks[i].getPositions(retainedPositions, offset, length);
+        }
+        return wrapBlocksWithoutCopy(length, blocks);
+    }
+
+    public Page copyPositions(int[] retainedPositions, int offset, int length)
+    {
+        requireNonNull(retainedPositions, "retainedPositions is null");
+
+        Block[] blocks = new Block[this.blocks.length];
+        for (int i = 0; i < blocks.length; i++) {
+            blocks[i] = this.blocks[i].copyPositions(retainedPositions, offset, length);
+        }
+        return wrapBlocksWithoutCopy(length, blocks);
+    }
+
+    public Page extractChannel(int channel)
+    {
+        return wrapBlocksWithoutCopy(positionCount, new Block[] {this.blocks[channel]});
+    }
+
+    public Page extractChannels(int[] channels)
+    {
+        requireNonNull(channels, "channels is null");
+
+        Block[] blocks = new Block[channels.length];
+        for (int i = 0; i < channels.length; i++) {
+            blocks[i] = this.blocks[channels[i]];
+        }
+        return wrapBlocksWithoutCopy(positionCount, blocks);
+    }
+
+    public Page prependColumn(Block column)
+    {
+        if (column.getPositionCount() != positionCount) {
+            throw new IllegalArgumentException(String.format("Column does not have same position count (%s) as page (%s)", column.getPositionCount(), positionCount));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: &#39;getPositionCount&#39; expression, positionCount. This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 397. Consider formatting this data with &#39;%d&#39; instead.